File: leetcode/1easy/위성원/1656.ts
class OrderedStream {
  stream: string[];

  constructor(n: number) {
    this.stream = new Array(n + 1);
  }

  getChunk(idKey: number, value: string): string[] {
    const chunk = [value];

    for (let i = idKey + 1; i < this.stream.length; i++) {
      const nextValue = this.stream[i];
      if (!nextValue) {
        break;
      }

      chunk.push(nextValue);
    }

    return chunk;
  }

  insert(idKey: number, value: string): string[] {
    this.stream[idKey] = value;

    // 왼쪽에 끊어진게 없는지 확인
    let flag = true;
    let j = idKey - 1;
    while (j > 0) {
      if (!this.stream[j]) {
        flag = false;
        break;
      }
      j--;
    }

    if (flag) {
      return this.getChunk(idKey, value);
    }

    return [];
  }
}

/**
 * Your OrderedStream object will be instantiated and called as such:
 * var obj = new OrderedStream(n)
 * var param_1 = obj.insert(idKey,value)
 */

File: leetcode/1easy/이경호/1656.cpp
#include<vector>
#include<string>

using namespace std;

class OrderedStream {
private:
  vector <string> stream;
  int pointer;
public:
  OrderedStream(int n) {
    stream.resize(n);
    pointer = 0;
  }

  vector<string> insert(int idKey, string value) {
    stream[idKey - 1] = value;
    vector<string> result;
    while (pointer < stream.size() && !stream[pointer].empty())
    {
      result.push_back(stream[pointer]);
      pointer++;
    }
    return result;
  }
};
